sstables/consumer.hh

```diff
@@ -585,10 +585,11 @@ public:
                 assert(data.size() == 0);
                 _remain -= orig_data_size;
                 if (skip.get_value() >= _remain) {
+                    skip_bytes skip_remaining(_remain);
```

Comment:
I didn't understand why the remaining to skip is _remain and not skip.get_value() - _remain()?

Reply:
We decrease `_remain` each time we consume a buffer (or its prefix) using `process()`, and when `_remain` reaches 0 we don't process any more buffers.
In the case that we are changing we've just consumed a buffer (updating `_remain -= orig_data_size`)  and received info about how many additional bytes after the consumed buffer we want to skip.
We only want to skip bytes that were going to be processed, so we skip only `_remain`.
The value of `skip.get_value() - _remain` would be the number of bytes that we weren't going to process anyway

Reply:
I'm not sure that explains things, so I will add the regression test for it

Reply:
I see. I think I understand now (I misunderstood what _remain was). Yes, a test would be really useful to convince people that it's a real bug - and also useful in the future if someone rewrites this code again and introduces bugs again.

Reply:
> Also, how did you test this? Can you reproduce the bug?

I've actually encountered this problem when working on https://github.com/scylladb/scylla/issues/2388. 
After my fix to it, `index_consume_entry_context` was not working correctly (boost/sstable_3_x_test was failing), but it started working after I disabled `options.dynamic_adjustments` of the underlying stream.
The actual reason it failed, was that with `dynamic_adjustments`, we may change the size of buffers that the input_stream is returning, and with smaller buffers we managed to hit this case.

Reply:
It wasn't detected before, probably because we only return `skip_bytes` in 2 cases; one of which we never fast forward anyway(index_reader), and one where we know that we have a next partition to read in mx sstable reader.